src/jdk.jdeps/share/classes/com/sun/tools/jnativescan/NativeMethodFinder.java

```diff
@@ -0,0 +1,144 @@
+import java.lang.classfile.constantpool.InterfaceMethodRefEntry;
+import java.lang.classfile.constantpool.MemberRefEntry;
+import java.lang.classfile.constantpool.MethodRefEntry;
+import java.lang.classfile.instruction.InvokeInstruction;
+import java.lang.constant.ClassDesc;
+import java.lang.constant.MethodTypeDesc;
+import java.lang.reflect.AccessFlag;
+import java.util.*;
+
+class NativeMethodFinder {
+
+    // ct.sym uses this fake name for the restricted annotation instead
+    // see make/langtools/src/classes/build/tools/symbolgenerator/CreateSymbols.java
+    private static final String RESTRICTED_NAME = "Ljdk/internal/javac/Restricted+Annotation;";
+
+    private final Map<MethodRef, Boolean> CACHE = new HashMap<>();
+    private final ClassResolver classesToScan;
+    private final ClassResolver systemClassResolver;
+
+    private NativeMethodFinder(ClassResolver classesToScan, ClassResolver systemClassResolver) {
+        this.classesToScan = classesToScan;
+        this.systemClassResolver = systemClassResolver;
+    }
+
+    public static NativeMethodFinder create(ClassResolver classesToScan, ClassResolver systemClassResolver) throws JNativeScanFatalError, IOException {
+        return new NativeMethodFinder(classesToScan, systemClassResolver);
+    }
+
+    public Map<ScannedModule, Map<ClassDesc, List<RestrictedUse>>> findAll() throws JNativeScanFatalError {
+        Map<ScannedModule, Map<ClassDesc, List<RestrictedUse>>> restrictedMethods = new HashMap<>();
+        classesToScan.forEach((_, info) -> {
+            ClassModel classModel = info.model();
+            List<RestrictedUse> perClass = new ArrayList<>();
+            classModel.methods().forEach(methodModel -> {
+                if (methodModel.flags().has(AccessFlag.NATIVE)) {
+                    perClass.add(new NativeMethodDecl(MethodRef.ofModel(methodModel)));
+                } else {
+                    Set<MethodRef> perMethod = new HashSet<>();
+                    methodModel.code().ifPresent(code -> {
+                         try {
+                             code.forEach(e -> {
+                                 switch (e) {
+                                     case InvokeInstruction invoke -> {
+                                         MethodRef ref = MethodRef.ofMemberRefEntry(invoke.method());
+                                         if (isRestrictedMethod(ref)) {
+                                             perMethod.add(ref);
+                                         }
+                                     }
+                                     default -> {
+                                     }
+                                 }
+                             });
+                         } catch (JNativeScanFatalError e) {
+                             throw new JNativeScanFatalError("Error while processing method: " +
+                                     MethodRef.ofModel(methodModel), e);
+                         }
```

Comment:
I'm re-wrapping the exception here, thrown from `SystemClassResolver::lookup`, in order to include the method that called the missing system class in the error message as well. (See also the code added in `Main` to print out an exception cause if it is a `JNativeScanFatalError`).